src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/WebRequestSession.cs

```diff
@@ -13,6 +14,20 @@ namespace Microsoft.PowerShell.Commands
     /// </summary>
     public class WebRequestSession
     {
+        private bool _changed;
+        private CookieContainer _cookies;
+        private bool _useDefaultCredentials;
+        private ICredentials _credentials;
+        private X509CertificateCollection _certificates;
+        private IWebProxy _proxy;
+        private int _maximumRedirection;
+        private HttpClientHandler _handler;
```

Comment:
The values in httpclienthandler are processed versions of the ones in the websession. We would still need some way of comparing what has been set with what we are needing to change because even if we set the same value on the HttpClientHandler that we've already set, it will still throw if it's already done a request. The setters all throw immediately if a request has already been issued, they don't do any kind of comparison with the value you're trying to set and the value they already have.

Reply:
Yes, we need the checks. I say we have no need to duplicate - we can get properties from target object. If any check fails we will recreate target object.

Reply:
This would save some memory in the WebRequestSession used to cache the settings by adding considerable additional conditionals to the path to setup the handler in the WebRequestPSCmdlet.Common.cs.

There is not a one-one mapping between the parameters to IWR and what is set on the handler in all cases, so we'd need to either duplicate all that logic in a validation phase, or put it inline in the code that creates it, with the need to then double back and start again once anything fails the validation. The additional checks will increase the number of conditionals in the cmdlet which leads to increased complexity and difficulty to maintain. I don't think this will make the code simpler or easier to understand, but perhaps I'm misunderstanding what you are suggesting.

Reply:
We have 6 scenario:
1. Use WebSession only internally. (The cmdlet must dispose the WebSession.)
2. The cmdlet issues the WebSession and it is reused as-is
3. The cmdlet issues the WebSession and it is modified by users before reuse.
4. User creates the WebSession and it is reused as-is
5. User creates the WebSession, uses in cmdlet and it is reused as-is
6. User creates the WebSession, uses in cmdlet, modify and reused

We use WebSession as argument store. So I'd expect all logic is in WebSession.

Reply:
Perhaps I'm misunderstanding your change proposal, since I agree with your 6 scenarios above and the patch is using the WebSession parameter as an argument store. Apart from the `HttpClientHandler _handler`, all those properties you listed (repeated below) already exist in the `WebRequestSession` class. The patch converts them to non-auto properties so the setter can detect if anyone is altering them through the public APIs and/or IWR parameters. I also added a few other properties that alter the properties of the `HttpRequestHandler` so I can determine if they have also been changed by the user at the time of invocation.

```csharp
        private X509CertificateCollection _certificates;
        private IWebProxy _proxy;
        private int _maximumRedirection;
        private HttpClientHandler _handler;
```

Reply:
Okay, let's see how your implementation works and make improvements if necessary.